### ISSUE TITLE
Colorbar on 3D interactive

### ISSUE THREAD

**Issue by RibeiroAndre** [JuliaPlots/Makie.jl#856]
Hi!
I'm working on something like this tutorial: http://juliaplots.org/MakieReferenceImages/gallery//fem_mesh_3d/index.html and I can't seem to add a Colorbar. Looking at the other question on Colorbars on meshes didn't help me and the layout tutorial didn't help me either. I think 3D interactive must be slightly different.
Thanks a lot!

**Comment by jkrumbiegel** [comment]
This works for me (without cat though, should still work with cat)

```julia
f, _, m = mesh(Point2f0[(0, 0), (1, 0), (0, 1)], [1, 2, 3],
    color = 1:3, shading = false)
Colorbar(f[1, 2], m, width = 20)
f
```

<img width="418" alt="grafik" src="https://user-images.githubusercontent.com/22495855/109198432-808a0880-779e-11eb-9897-7d6285bf2cfc.png">

**Comment by jkrumbiegel** [comment]
or 3D

```julia
f, _, m = mesh(Point3f0[(0, 0, 0), (1, 0, 0), (0, 1, -1)], [1, 2, 3],
    color = 1:3, shading = false)
Colorbar(f[1, 2], m, width = 20)
f
```
<img width="420" alt="grafik" src="https://user-images.githubusercontent.com/22495855/109198577-ad3e2000-779e-11eb-8e80-85e5a27531c2.png">

**Comment by RibeiroAndre** [comment]
Wow, that easy, huh?
Can you explain the 3 values that mesh returns? "f" is the figure, I guess. What is "m" and the blank one?
Thanks a lot!

**Comment by jkrumbiegel** [comment]
Described here: http://makie.juliaplots.org/stable/plot_method_signatures.html#Non-Mutating

figure, axis, plot

**Comment by RibeiroAndre** [comment]
Cool, thanks! I was missing the plot part in my attempts!
Thanks again!

**Comment by RibeiroAndre** [comment]
Can I get some more help? =P
On that same example you made, how easy is it to add a slider that controls `colorrange`, a menu that changes the variable being colored on the triangle (say in your case you have `color=1:3` and `color2=2:4`, and a toggle that turns a `wireframe!` on and off?
I can't find any examples like that in 3D. They all create a figure first and I'm not sure how to do that with `mesh`.
Thanks a lot! Sorry if this is too much. Trying to make a basic post proc window =]

**Comment by ffreyer** [comment]
> They all create a figure first and I'm not sure how to do that with mesh.

It works the same way, you just want an `LScene` instead of an `Axis`.

```julia
fig = Figure()
s = LScene(fig[1, 1])
mesh!(s, ...)
```

> toggle that turns a wireframe! on and off?

You can do that via `wireframe!(s, ..., visble = toggle.active)`

For the rest you just need to convert `slider.value` and `menu.selection` to whatever you need and put that in as the value of the attribute like with toggle.

**Comment by RibeiroAndre** [comment]
That was incredibly helpful! It took all morning and a lot of iterations, but I made a little GUI with a menu that selects variables and a toggle for wireframe.
```
postFig = Figure()
postScene = LScene(postFig[1, 1]);
postToggle = Toggle(postFig,active=true);
postToggleLabel = Label(postFig, lift(x -> x ? "Mesh On" : "No Mesh", postToggle.active));
postMenu = Menu(postFig, options = [("Cp",pointsCp), ("Vmag",pointsVmag)], i_selected=1, selection=pointsCp, width = 100);
postFig[1, 2] = grid!([postToggleLabel postToggle;Label(postFig,"Variable") postMenu], tellheight = false)
AbstractPlotting.wireframe!(postScene,surfaceMesh, visible = postToggle.active)

minval = minimum(pointsCp);
maxval = maximum(pointsCp);
postSliderMin=Slider(postFig[2,1],range=minval:(maxval-minval)/100:maxval,startvalue=minval)
postFig[2,2]=Label(postFig,"Min")
postSliderMax=Slider(postFig[3,1],range=minval:(maxval-minval)/100:maxval,startvalue=maxval)
postFig[3,2]=Label(postFig,"Max")

postMesh=AbstractPlotting.mesh!(postScene,surfaceMesh,color=menu.selection[],colormap=:turbo,colorrange=(postSliderMin.value,postSliderMax.value))
AbstractPlotting.Colorbar(postFig[1, 3],postScene.scene.plots[3], width = 20)

on(postMenu.selection) do s
    postMesh.color = s;
	minval = minimum(s);
	maxval = maximum(s);
	postSliderMin.range=minval:(maxval-minval)/100:maxval;
	set_close_to!(postSliderMin)=minval
	postSliderMax.range=minval:(maxval-minval)/100:maxval;
	set_close_to!(postSliderMax)=maxval
	postMesh.colorrange=(minval,maxval);
end
```
Everything works, except for the slider. It seems `colorrange=(postSliderMin.value,postSliderMax.value)` doesn't link the colorrange to the slider value, just sets it to the current value. So when I move the slider nothing happens.
I also tried `lift(x->x,postSliderMin.value`, but that didn't work either.
I think this is my last question, so if you guys can help, I'll stop annoying you =]
Thanks a lot!

**Comment by ffreyer** [comment]
That should be `colorrange = lift((x, y) -> (x, y), postSliderMin.value, postSliderMax.value)`. That way you create a new `Node` that holds a tuple and updates when either source updates, rather than a tuple that holds two nodes.

**Comment by RibeiroAndre** [comment]
Wow, I would have never thought of that =P. Thanks a lot!
Now it kind of works, but when I use the menu I still have some issues. The colors don't update until I touch the max slider. If I touch the min slider, I get `ArgumentError: reducing over an empty collection is not allowed`. I suspect the `on` loop has an issue. Is there a specific way I should reassign a slider range? (Like the specific way I set the value).
Thanks again!

**Comment by jkrumbiegel** [comment]
you can also do `@lift(($(postSliderMin.value), $(postSliderMax.value)))`, which in this case has a lot of parentheses maybe, but in general it can save some typing

**Comment by RibeiroAndre** [comment]
Ah, cool, thanks.
Ok, I clearly messed up `set_close_to!(postSliderMax)=maxval`. It should be `set_close_to!(postSliderMax,maxval)`. But that doesn't work either. Apparently I'm reducing over an empty collection?

**Comment by jkrumbiegel** [comment]
We can't run your code because there are many variables missing, but reducing over empty collection happens usually with `sum` on an empty vector or something like that

**Comment by RibeiroAndre** [comment]
I got it. The problem is that setting the range is tricky. Say the range was -2:-1 when I was showing a variable. Then say the range of a new variable is 0:1. If I set the new range and value for SliderMin, it is now larger than the range and value of SliderMax, so the mesh messes up the color map (since it is now 0:-1, which AbstractPlotting doesn't seem to like).
So something stupid like this:
```
on(postMenu.selection) do s
    postMesh.color = s;
	@show minval = minimum(s);
	@show maxval = maximum(s);
	postSliderMin.range=-1e9:100:-1e8;
	set_close_to!(postSliderMin,-1e9);
	postSliderMax.range=1e8:100:1e9;
	set_close_to!(postSliderMax,1e9);
	postSliderMin.range=minval:(maxval-minval)/100:maxval;
	set_close_to!(postSliderMin,minval);
	postSliderMax.range=minval:(maxval-minval)/100:maxval;
	set_close_to!(postSliderMax,maxval);
end
```
Fixes the problem. And I have to be careful not to cross the sliders...
If anyone has a more elegant solution (I'd love to somehow lock the sliders so the don't cross, but I couldn't figure that out), I'm all for it. If not, I really appreciate the help! Pretty happy with the outcome =D
